src/webui/webapplication.cpp

```diff
@@ -914,76 +804,97 @@ void WebApplication::action_command_addCategory()
-    action_ = DEFAULT_ACTION;
-
-    parsePath();
-
-    if (args_.contains(".") || args_.contains("..")) {
+    const QStringList pathItems {request().path.split('/', QString::SkipEmptyParts)};
+    if (pathItems.contains(".") || pathItems.contains("..")) {
         qDebug() << Q_FUNC_INFO << "Invalid path:" << request().path;
         status(404, "Not Found");
         return;
     }
 
-    if (!isPublicScope() && !sessionActive()) {
-        status(403, "Forbidden");
-        return;
-    }
+    if (request().path.startsWith(PATH_PREFIX_API)) {
+        const QString methodName {request().path.mid(PATH_PREFIX_API.size())};
+        if (!isPublicAPI(methodName) && !sessionActive()) {
+            status(403, "Forbidden");
+            return;
+        }
 
-    if (actions_.value(scope_).value(action_) != 0) {
-        (this->*(actions_[scope_][action_]))();
+        if (!QMetaObject::invokeMethod(this, methodName.toLatin1().constData())) {
+            status(404, "Not Found");
+            qDebug() << Q_FUNC_INFO << "Resource not found:" << request().path;
+        }
     }
     else {
-        status(404, "Not Found");
-        qDebug() << Q_FUNC_INFO << "Resource not found:" << request().path;
+        getWebUIFile();
     }
 }
 
-void WebApplication::parsePath()
+void WebApplication::configure()
+{
+    const auto pref = Preferences::instance();
+    m_username = pref->getWebUiUsername();
+    m_password = pref->getWebUiPassword();
+
+    const QString rootFolder = Utils::Fs::expandPathAbs(
+                !pref->isAltWebUiEnabled() ? WWW_FOLDER : pref->getWebUiRootFolder());
+    if (rootFolder != m_rootFolder)
+        m_translatedFiles.clear();
+    m_rootFolder = rootFolder;
+}
+
+bool WebApplication::readFile(const QString &path, QByteArray &data, QString &type)
 {
-    if (request().path == "/") action_ = WEBUI_ACTION;
+    type = getContentType(path);
+    const bool isTranslatable {(type == Http::CONTENT_TYPE_HTML) || (type == Http::CONTENT_TYPE_JS)};
+    const QDateTime lastModified {QFileInfo(path).lastModified()};
 
-    // check action for requested path
-    QStringList pathItems = request().path.split('/', QString::SkipEmptyParts);
-    if (!pathItems.empty() && actions_.contains(pathItems.front())) {
-        scope_ = pathItems.front();
-        pathItems.pop_front();
+    // find translated file in cache
+    if (isTranslatable) {
+        auto it = m_translatedFiles.constFind(path);
+        if ((it != m_translatedFiles.constEnd()) && (lastModified <= (*it).lastModified)) {
+            data = (*it).data;
+            return true;
+        }
     }
 
-    if (!pathItems.empty() && actions_[scope_].contains(pathItems.front())) {
-        action_ = pathItems.front();
-        pathItems.pop_front();
+    QFile file(path);
+    if (!file.open(QIODevice::ReadOnly)) {
+        qDebug("File %s was not found!", qUtf8Printable(path));
+        return false;
     }
 
-    args_ = pathItems;
-}
+    data = file.readAll();
```

Comment:
upper limit please.

Reply:
We didn't have it later. What's the purpose? Shouldn't user control itself what it have in its www root, IMO.

Reply:
>What's the purpose? 

It's reading like: hey I'm going to read a file and I don't care how big it is and I want it all in the memory NOW.
It's a bad practice for an app to grant users unlimited resources.

Reply:
But what is a reasonable limit? And what status to return if you exceed this size?

Reply:
Well, now limited to 1MB and send 500 (Internal server error) when exceeded. Waiting for your suggestions.

Reply:
Thank you! I was considering 1MB or 10MB, either is fine.